File: app/javascript/queries/activity.js
import gql from 'graphql-tag'

export default gql`
  query Activity($year: Int!, $type: ActivityType!, $slug: String!) {
    activity(year: $year, type: $type, slug: $slug) {
      id
      name
      type
      slug
      url
      description

      image {
        name
        thumbnail
        small
        medium
        full
      }

      ...on Workshop {
        levels
      }

      associated {
        id
        name
        type
        url

        sessions {
          startsAt
        }
      }

      festival {
        year
      }

      presenters {
        id
        name
        origin
        bio
        image {
          thumbnail
          small
        }
      }

      sessions {
        id
        startsAt
        endsAt

        activity {
          id
        }

        venue {
          id
          name
          address
        }
      }
    }
  }
`

File: app/graphql/types/activity.rb
module Types
  module Activity
    include Types::BaseInterface

    field :id, ID, null: false
    field :name, String, null: false
    field :type, ActivityType, null: false
    field :slug, String, null: false
    field :description, String, null: true
    field :festival, Types::Festival, null: false
    field :url, String, null: false
    field :associated, [Types::Activity], null: false
    field :sessions, [Types::Session], null: false
    field :presenters, [Types::User], null: false

    def resolve_type(object, context)
      case object
      when Workshop then Types::Workshop.graphql_definition
      when Show then Types::Show.graphql_definition
      else
        raise "Unexpected object: #{object.inspect}"
      end
    end

    def presenters
      object.presenters.includes(:user).map(&:user)
    end
  end
end
